[SYCL][Joint Matrix Spec] Add new API for out of bounds fill/load/store

dkhaldi opened the pull request:
Code example to show usage can be found here:
https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Matrix/joint_matrix_out_bounds.cpp

YuriPlyakhin commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -198,6 +198,67 @@ joint_matrix_apply(sg, A, [=](T &val, size_t row, size_t  col) {
+
+=== Load/Store/Fill With Out-of-Bounds Checks
+`joint_matrix_load`, `joint_matrix_store`, and `joint_matrix_fill`
+operations do not do bounds checking. When the global matrix size does
+not multiply `Rows` and `Cols` of the joint matrix, this extension
```
```suggestion
operations do not do bounds checking. When the global matrix size is
not multiple of the joint matrix size, this extension
```

YuriPlyakhin commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -198,6 +198,67 @@ joint_matrix_apply(sg, A, [=](T &val, size_t row, size_t  col) {
+checking the bounds. For load and fill, the out-of-bounds elements are
+set to 0. For the store, they are dropped out.
+
+These functions are similar to the existing ones but they
```
```suggestion
These functions are similar to the ones without bounds checking, but they
```

dkhaldi commented:
@gmlueck, gentle ping to review.

dkhaldi commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -203,6 +203,96 @@ In the case of `ext_intel_packed` matrix memory layout, `row` and
 `col` represent the coordinates in the logical matrix before VNNI
 transformation.
 
+=== Load/Store/Fill With Out-of-Bounds Checks
```
The check is done compared to the big matrix size comparing the address to be loaded (ptr) with the big matrix size (Height). If ptr > Height, the element should not be loaded.
if the remainder is passed, how is the implementation going to compare with it?

gmlueck replied:
With your recent change that adds `coordX` and `coordY` parameters, I see how this could work for the load and store APIs.  However, I still don't see how the fill API can work.

This API structure seems confusing to me, though, and probably less efficient.  I suspect users will get confused that the `src` pointer to `joint_matrix_load` is completely different from the `src_base` pointer to `joint_matrix_load_checked`.  It seems like it would be less confusing and more efficient to have the user pass values for `Width` and `Height` that are the *remaining* amount of memory in the big matrix.  Maybe this would be easier to explain in a meeting where we can draw pictures?

dkhaldi replied:
Yes, let's discuss fill and remainder approach in a meeting.

dkhaldi commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
````diff
@@ -203,6 +203,106 @@ In the case of `ext_intel_packed` matrix memory layout, `row` and
+
+```c++
+namespace sycl::ext::intel::experimental::matrix {
+
````
@fda0 does this change address the missing arguments to generate efficiently 2d load/store in IGC?

dkhaldi replied:
Yes I changed it to add the two offset coordinates separately and excluded them from the base pointer calculation.

dkhaldi commented:
Notes after online discussion with @gmlueck:
- Annotated ptr load/store overloads needs to be in oneapi namespace for use on targets that support cache control hints
- Prefetch needs to be in oneapi namespace 
- Out of bounds support:
	- Do we need emulation when hardware cannot do the bounds check?
        - Do we need these for AMX?
        - Fill does not need base pointer but will need coordinates and stride to be added
        -Should we add base ptr, coord x and y to the current joint_matrix_load/store to match the checked variants?

dkhaldi commented:
@gmlueck, I changed the API to stick with offset pointer, no separate coordinates are needed. We did that because we found that alignment restriction does not really make a difference whether we use offset or base pointer (4 bytes alignments must be honored regardless).

However, we found an another issue if we stick with offset pointer which is the inability to support negative coordinates. There are use cases where users need to read backwards in memory and the extra elements (backward before the start of the memory pointer) need to be zeroed. If we stick with offset pointer (remainder Width and Height cannot express the negative coordinates case), we will have to restrict the use of these API to only positive coordinates which can be limiting to the real use cases.

uphoffc commented:
Here's another thing to consider.

Assume that I tile my row-major MxN matrix into 8x16 tiles. Using the current proposal, writing a loop accessing every tile looks like the following:
1)
```c++
for (m = 0; m < M; m += 8) {
  for (n = 0; n < N; n += 16) {
    joint_matrix_load_checked(sub_group, result, base_ptr + n + m * Stride, Stride, M - m, N - n);
  }
}
```
In contrast, with base_ptr + coordinate we have
2)
```c++
for (m = 0; m < M; m += 8) {
  for (n = 0; n < N; n += 16) {
    joint_matrix_load_checked(sub_group, result, base_ptr, Stride, M, N, int2{m, n});
  }
}
```

In variant 1) we have additional arithmetic going on; we need 4 add / subs and 1 mul, that would otherwise be done in hardware. Moreover, we have to do the offset calculation ourselves every time and must not forget to adjust the bounds - imho much more error prone and complicated than 2).

dkhaldi commented:
@gmlueck, as we discussed on Monday, I put back the coords arguments and base pointer as they are needed for negative coordinates case. 
Can you please review/approve?

dkhaldi commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -203,6 +203,147 @@ In the case of `ext_intel_packed` matrix memory layout, `row` and
+          use Use, layout Layout, typename Tv>
+void joint_matrix_fill_checked(Group g, joint_matrix<Group, T, Use, Rows,
+          Cols, Layout> &m, Tv v, size_t Stride, size_t Height, size_t
+          Width, size_t coordX, size_t coordY);
```
@MrSidims, 
In the load/store, the check does the following (let's just consider one horizontal dimension to simplify):
if (base_ptr + coordY + Cols > base_ptr+Width) --> out of bounds

Since we intend  to do the check just like load and store (just considering base_ptr=0), we need to add these arguments: Height, Width, Stride, CoordX, and CoordY. The check will do the following:
if (coordY + Cols > Width) --> out of bounds

dkhaldi commented:
> LGTM with the assumption, that Stride is removed form matrilx_fill API

Yes, I just removed stride. So the expectation is the following WRT joint_matrix_fill_checked:

- Check for the y dimension (columns): coordy+smallmatrixcols >? width

- Check for the x dimension (rows): coordx + smallmatrixrows>? Height

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -203,6 +203,147 @@ In the case of `ext_intel_packed` matrix memory layout, `row` and
+layout::row_major> sub_b;` from `ptr + 9` checks that `ptr + 9` does
+not exceed `ptr + 10` where 10 is `Width` so it reads only 1 element
+for that iteration. The 2 columns left that exceed the total `10`
+columns are filled by zero.
```
I find this paragraph extremely confusing.  Some specific things that seem confusing:

* The phrase "for each sub-group execution" seems irrelevant.  You are describing a single joint matrix load operation here, so there is no need to talk about iterations in a larger algorithm.
* What are `src` and `ptr`?  None of the APIs have parameters with these names.
* It seems like `CoordX` and `CoordY` should appear somewhere in this description since they indicate the starting point of the joint matrix.

gmlueck commented:
I found this information in an email conversation we had back in January:

> There are two issues with this feature because we only plan on adding it on hardware that offers this capability, no emulation:
> * This will only be supported on PVC and FS1, not the CPU(AMX). So we should add a query value about OOB checking. 
> * They need to be aligned by 4 bytes because we want to avoid runtime checking. I will add a note about this requirement. We may need to add this as an extension to the OOB query value. Something like: 
>    * `bool info::device::matrix::out_of_bounds` --> support for OOB or not
>    * `size_t info::device::matrix::out_of_bounds_alignment` -->  alignment requirement 

I'm not sure if these limitations still exist.  If so, I think these queries need to be added to the spec.

Rather than adding `info::device::matrix::out_of_bounds`, we should add an aspect for this.  Maybe name it like `ext_intel_matrix_checked`, as this aligns with the names of the APIs (e.g. `joint_matrix_load_checked`).

None of the information descriptor queries in sycl_ext_oneapi_matrix have a namespace like `info::device::matrix::FOO`, so I don't think you should add that namespace just for this one query.  Maybe name it like `ext::intel::experimental::info::device::matrix_checked_alignment`.

dkhaldi commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -203,6 +203,147 @@ In the case of `ext_intel_packed` matrix memory layout, `row` and
+in the SYCL extenion
+link:../../proposed/sycl_ext_intel_cache_controls.asciidoc[sycl_ext_intel_cache_controls],
+and for the alignment property defined in the SYCL extension
+link:../../experimental/sycl_ext_oneapi_kernel_arg_properties.asciidoc[sycl_ext_oneapi_kernel_arg_properties].
```
I will remove alignment from the checked APIs because only cache control is what is being considered for implementation so far.
Alignments will be handled by the query like you suggested in your last comment. So users must make sure their pointers are 4 bytes aligned. Otherwise, results are undefined.

dkhaldi commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -203,6 +203,147 @@ In the case of `ext_intel_packed` matrix memory layout, `row` and
+`joint_matrix_store`. But they are different in three ways:
+
+* The pointer `base_src` or `base_dest` designates the base pointer of
+the original memory matrix excluding any offset calculations. The
```
Yes, I like this wording better, thanks!

dkhaldi commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -203,6 +203,147 @@ In the case of `ext_intel_packed` matrix memory layout, `row` and
+the original memory matrix excluding any offset calculations. The
+coordinates are given separately here.
+* The coordinates `CoordX` and `CoordY` to calculate the  pointer
+offset to load/store are given as separate arguments.
```
Correct

dkhaldi commented:
@gmlueck, I addressed your comments mainly:
- Add new aspect for the checked APIs `ext_intel_matrix_checked`
- Restructuring and more clarifications: add subsections for aspect, introduction, API, restrictions and query 
- Add Device Information Descriptors for the checked APIs restrictions

Can you please take a look again?

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -340,9 +384,35 @@ void joint_matrix_store_checked(Group g,
+
+- For 8 bits data type, `Width` must be a multiple of 4, For 16 bits
+data type, `Width` must be a multiple of 2. So `Width` must be a
+multiple of 4 divided by size of the element type (`4/sizeof(T)`).
```
Describing it this way is confusing, and not really right.  For example, you first say that the base pointer must be 4 byte aligned, and then you say there is a query to get this alignment.  Presumably, we expect the alignment requirement to be different someday, so the spec should not say the requirement is 4 bytes.  I think you should remove these bullet points and say something like:

> Applications must adhere to certain alignment restrictions when using the checked APIs described in this section.  This extension provides the following queries to get these requirements:
>
> [Table]

The "Description" column will need to be expanded a bit to describe some of the details that are in the bullet points.  For example, the description of `matrix_checked_alignment` might say:

> Tells the required alignment (in bytes) of the base pointer for `joint_matrix_load_checked` and `joint_matrix_store_checked`.

If you want to document the alignment requirements of our current hardware, you could do that after the table describing the queries.  But you should somehow make it clear that the values only apply to our existing devices, and future devices may have different requirements.

dkhaldi replied:
That's a good point, this is tying the API to a specific hardware which should be avoided.
yes, I will add the alignment requirements as an appendix like we did for the combinations table per each device. Here, these only apply to PVC for now.

dkhaldi commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -244,16 +249,42 @@ the out of bounds data (depicted as x here).
 |<------------`Stride`---------->|
```
>> Using a horizontal line for "height" is confusing. 

This is not a horizontal line for "height". It is just one of the rows as "...". But I agree with the confusion. I will improve it.

>> "Stride" is normally a power of two?

There is no restriction on stride

I will match the numbers in the example with the picture

dkhaldi commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -244,16 +249,42 @@ the out of bounds data (depicted as x here).
+not multiple of the joint matrix size, this extension adds a new form
+of API to load, store, and fill joint matrix while checking the
+bounds. For load and fill, the out-of-bounds elements are set to
+0. For the store, they are dropped out.
```
Your description looks better, I will change to it. 
However, I don't prefer to use the word "padding" as this may suggest padding occurred (by the user or compiler) which is not necessarily the case, these data are just there in memory. So I would use the word "extra".

yes, the invalid address description is correct as these functions do not try to access the extra elements.

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -203,6 +203,260 @@ In the case of `ext_intel_packed` matrix memory layout, `row` and
+data type, `RowIndex` must be a multiple of 2. So `RowIndex` must be a
+multiple of 4 divided by size of the element type (`4/sizeof(T)`).
+
+- For 8 bits data type, `GlobalCols` must be a multiple of 4, For 16 bits
```
```suggestion
- For 8 bits data type, `GlobalCols` must be a multiple of 4. For 16 bits
```

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc:
```diff
@@ -203,6 +203,260 @@ In the case of `ext_intel_packed` matrix memory layout, `row` and
+|Tells the required alignment (in bytes) of the base pointer for
+`joint_matrix_load_checked` and `joint_matrix_store_checked`.
+|`ext::intel::experimental::info::device::matrix_checked_rowindex_multiple_of<T>`|
+`size_t`|Returns a value, of which `RowIndex` must be multiple of.
```
Kind of a nit, but it would be good to add a sentence:

> Where `T` is the element type of the matrix.

However, this makes me wonder how this works for a matrix that uses TF32.  Should `T` be `precision::tf32` or `float` (which is the storage type when using TF32)?

dkhaldi replied:
We should be able to use the element type `precision::tf32`

gmlueck replied:
I'd suggest adding another sentence to make this clear.  Something like:

> When using the matrices with the machine learning types, `T` should be the element type (e.g. `precision::tf32`) not the storage type.

dkhaldi commented:
@intel/llvm-gatekeepers, please help merge.
